skimage/metrics/tests/test_structural_similarity.py

```diff
@@ -204,42 +207,64 @@ def test_gaussian_structural_similarity_vs_IPOL():
     https://github.com/scikit-image/scikit-image/pull/4913#issuecomment-700653165
     """
     mssim_IPOL = 0.357959091663361
+    assert cam.dtype == np.uint8
+    assert cam_noisy.dtype == np.uint8
     mssim = structural_similarity(cam, cam_noisy, gaussian_weights=True,
                                   use_sample_covariance=False)
     assert_almost_equal(mssim, mssim_IPOL, decimal=3)
 
 
-def test_mssim_vs_legacy():
+@pytest.mark.parametrize(
+    'dtype', [np.uint8, np.int32, np.float16, np.float32, np.float64]
+)
+def test_mssim_vs_legacy(dtype):
     # check that ssim with default options matches skimage 0.17 result
     mssim_skimage_0pt17 = 0.3674518327910367
-    mssim = structural_similarity(cam, cam_noisy)
+    assert cam.dtype == np.uint8
+    assert cam_noisy.dtype == np.uint8
+    mssim = structural_similarity(cam.astype(dtype),
+                                  cam_noisy.astype(dtype), data_range=255)
     assert_almost_equal(mssim, mssim_skimage_0pt17)
 
 
-def test_mssim_mixed_dtype():
+def test_ssim_warns_about_data_range():
     mssim = structural_similarity(cam, cam_noisy)
-    with expected_warnings(['Inputs have mismatched dtype']):
-        mssim_mixed = structural_similarity(cam, cam_noisy.astype(np.float32))
-    assert_almost_equal(mssim, mssim_mixed)
+    with expected_warnings(['Setting data_range based on im1.dtype']):
+        mssim_uint16 = structural_similarity(cam.astype(np.uint16),
+                                             cam_noisy.astype(np.uint16))
+    with expected_warnings(['Setting data_range based on im1.dtype',
+                            'Inputs have mismatched dtype']):
+        mssim_mixed = structural_similarity(cam, cam_noisy.astype(np.int32))
 
     # no warning when user supplies data_range
     mssim_mixed = structural_similarity(
         cam, cam_noisy.astype(np.float32), data_range=255)
+
     assert_almost_equal(mssim, mssim_mixed)
 
+    # This is a wrong value, based on the fact that the data type led
+    # to infer the wrong data_range. But we issued a warning.
```

Comment:
```suggestion
    # The value computed for mssim_uint16 is wrong, because the
    # dtype of im1 led to infer an erroneous data_range. The user
    # is getting a warning about avoiding mistakes.
```
(rewording just for better understanding)